### ISSUE TITLE
Cannot backprop through `nn.BatchNorm` in `eval` mode

### ISSUE THREAD

**Issue by dedeswim** [pytorch/xla#3361]
## 🐛 Bug

If I try to backdrop through an `nn.BatchNorm` layer which is part of a model on which I have called `.eval()` my script crashes.

## To Reproduce

1. Create a model with a BatchNorm layer
2. Call `.eval()` on the model
3. Compute some loss
4. Backprop through the model with `autograd.grad(loss, x)` or with `loss.backward()` to compute the gradient w.r.t. the input `x`.

A Colab notebook based on the template to reproduce the issue can be found [here](https://gist.github.com/dedeswim/1e0c66a74c2eacba33a88259c1912309).

## Expected behavior

I should be able to backdrop through a model using BatchNorm even if it is in `.eval()` mode, which I can do when I run on GPU without using XLA.

## Environment

 - Reproducible on XLA backend [CPU/TPU]: TPU both on Colab and on a TPU VM
 - torch_xla version: 1.9

**Comment by JackCaoG** [comment]
Thanks @dedeswim for reporting the issue! I will try to repo it locally.

**Comment by dedeswim** [comment]
Hi @JackCaoG, by chance, did you manage to reproduce this locally? Please, let me know if you need more information to do so 😊

**Comment by JackCaoG** [comment]
Will try to do that today

**Comment by JackCaoG** [comment]
The good news is that we almost finished our colab wheel issue so when we fix the issue you will get it working on the colab. The bad news is I have yet to have time rootcausing this issue.. We are in the process of preparing for 1.11 release so my hand is a bit tight.. I will keep you updated.

**Comment by dedeswim** [comment]
Thanks a lot for the update. I'd be happy to help in case there is something I can do

**Comment by JackCaoG** [comment]
Ok I find the issue. Basically pt/xla was complaining about `mean` passed to  the `native_batch_norm_backward` has different dimension than the `feature`. In your example, feature dimension is 3 and mean was 0.

The cause of this mismatched is actually 
```
f.eval()
```

In `XLANativeFunctions::native_batch_norm`  a running mean was passed to pytorch/xla and we will update this running mean and return it to the caller. 
https://github.com/pytorch/xla/blob/cf2763033d1051dfe96984f44516edd50b444dd6/torch_xla/csrc/aten_xla_type.cpp#L2281

This `running_mean` eventually got passed to the `XLANativeFunctions::native_batch_norm_backward` as `save_mean` 
https://github.com/pytorch/xla/blob/cf2763033d1051dfe96984f44516edd50b444dd6/torch_xla/csrc/aten_xla_type.cpp#L2306

What pytorch/xla does today is when `native_batch_norm` is called with `training=false` (inference mode), we will drop the `runnign_mean` and return a 0.
https://github.com/pytorch/xla/blob/cf2763033d1051dfe96984f44516edd50b444dd6/torch_xla/csrc/tensor_methods.cpp#L2040

This 0 replaced the `running_mean` and when passed to the `native_batch_norm_backward` caused a shape mismatch. The natural solution seems to be not overwriting the `running_mean` but I am a bit unsure what's the expect behavior from the pytorch side. @mruberry Can you shed some light regarding this?

**Comment by JackCaoG** [comment]
@mruberry @bdhirsh Could you give me some guidance regarding this behavior?

**Comment by bdhirsh** [comment]
@JackCaoG I compared the CPU and XLA behavior - in that code snippet for XLA batch norm ([link](https://github.com/pytorch/xla/blob/cf2763033d1051dfe96984f44516edd50b444dd6/torch_xla/csrc/tensor_methods.cpp#L2040)), where is it actually overwriting `running_mean`? It looks like it takes in a `running_mean` tensor, and (without modifying the input), it returns a new zero-sized `mean` tensor.

The derivative formula for `native_batch_norm` ([here](https://github.com/pytorch/pytorch/blob/95204c4e2b30bda85d8907eb47763c05af8b474c/tools/autograd/derivatives.yaml#L1083)) just directly passes in the original `running_mean` input, which should still be size=3 in the example (since it was never modified).

Let me know if I'm missing something. For context, it looks like the CPU implementation does the same thing - when `train=false`, we return a new mean that is zero-sized, but that's fine because what gets passed to the backward is the original `running_mean` input (which has the correct size). ([here](https://github.com/pytorch/pytorch/blob/197764b35db68dee21ccc542d88a338f813623e7/aten/src/ATen/native/Normalization.cpp#L648))

**Comment by JackCaoG** [comment]
`but that's fine because what gets passed to the backward is the original running_mean input (which has the correct size`
I felt like this is not the case for pytorch/xla, but I need to verify.

**Comment by bdhirsh** [comment]
> `but that's fine because what gets passed to the backward is the original running_mean input (which has the correct size` I felt like this is not the case for pytorch/xla, but I need to verify.

Sounds good. I think I would expect it to be the case for XLA though... since that is the "default" derivative formula for `native_batch_norm` (defined in [core](https://github.com/pytorch/pytorch/blob/95204c4e2b30bda85d8907eb47763c05af8b474c/tools/autograd/derivatives.yaml#L1083)), meaning that all backends would use that same derivative formula by default. And I didn't see any place where pytorch/xla appears to override it (you only override `max_pool2/3d` afaik)

**Comment by JackCaoG** [comment]
I checked the pytorch implementation and find that `save_mean` is a `CPUFloatType{0}` if `BatchNorm` is only run in the eval mode. After `BatchNorm` being run in the `train` mode, it will be updated to the same as `running_mean`.

when `BatchNorm(train=false)` 
```
in backward running mean =  0
 0
 0
[ CPUFloatType{3} ]
in backward save mean = [ CPUFloatType{0} ]
```

when `BatchNorm(train=true)`
```
in backward running mean = 0.01 *
-1.5385
-8.1150
-2.7137
[ CPUFloatType{3} ]
in backward save mean = -0.1538
-0.8115
-0.2714
[ CPUFloatType{3} ]
```

There is a shape mismatched between `running_mean` and `saved_mean` when only eval mode is used.. Let me find a way to handle this corner case.

**Comment by JackCaoG** [comment]
hmm, so every time `eval` mode is enabled, `save_mean` will become a `CPUFloatType{0}` and pt/xla uses `save_mean` to compute the backward pass. Checking pytorch implementation to see what's the right thing to do here.

**Comment by JackCaoG** [comment]
I will follow pytorch implementation in https://github.com/pytorch/pytorch/blob/5dbec7c07c5eedd748fd56359c2d1b980dfa1037/aten/src/ATen/native/cpu/batch_norm_kernel.cpp#L59 to fix the backward pass when `train==false`

**Comment by JackCaoG** [comment]
I found out that this is not just a simple replacement, eval mode backward actually have a [different logic](https://github.com/pytorch/pytorch/blob/5dbec7c07c5eedd748fd56359c2d1b980dfa1037/aten/src/ATen/native/cpu/batch_norm_kernel.cpp#L413) compared to training. We currently only implement the logic to backward  the training `batchnorm` using `xla::BatchNormGrad`, I didn't find the corresponding xla function for inference backward(there is one for forward). I will check with the xla team.

**Comment by dedeswim** [comment]
This is not actually stale, it would be great if the issue got fixed! Thanks :blush:

**Comment by JackCaoG** [comment]
I think I know the issue but I don't have bandwidth to fix it now. I will ask around and see if I can find someone else to fix it.

**Comment by dedeswim** [comment]
Thanks a lot! :)

**Comment by YuanEldaif** [comment]
Is there any update on the nn.BatchNorm back prorogation issue in eval mode?  It works on gpu & cpu devices but doesn't work on TPU. Greatly appreciate it!
